packages/ember-glimmer/lib/component-managers/custom.ts

```diff
@@ -1,32 +1,119 @@
-import DefinitionState from './definition-state';
-
-export interface CustomComponentManagerDelegate<T> {
-  version: 'string';
-  create(options: { ComponentClass: T; args: {} }): T;
-  getContext(instance: T): Opaque;
-  update(instance: T, args: {}): void;
-  destroy?(instance: T): void;
-  didCreate?(instance: T): void;
-  didUpdate?(instance: T): void;
-  getView?(instance: T): any;
+
+const CAPABILITIES = {
+  dynamicLayout: false,
+  dynamicTag: false,
+  prepareArgs: false,
+  createArgs: true,
+  attributeHook: false,
+  elementHook: false,
+  createCaller: false,
+  dynamicScope: true,
+  updateHook: true,
+  createInstance: true,
+};
+
+export interface OptionalCapabilities {
+  asyncLifecycleCallbacks?: boolean;
+  destructor?: boolean;
 }
 
-export interface ComponentArguments<T = {}> {
+export function capabilities(
+  managerAPI: '3.4',
+  options: OptionalCapabilities = {}
+): ComponentCapabilityMask {
+  assert('Invalid component manager compatibility specified', managerAPI === '3.4');
+
+  return {
+    asyncLifeCycleCallbacks: !!options.asyncLifecycleCallbacks,
+    destructor: !!options.destructor,
+  };
+}
+
+export interface DefinitionState<ComponentInstance = Opaque> {
+  name: string;
+  ComponentClass: Factory<ComponentInstance>;
+  symbolTable: ProgramSymbolTable;
+  template?: any;
+}
+
+export interface ComponentCapabilityMask {
+  asyncLifeCycleCallbacks: boolean;
+  destructor: boolean;
+}
+
+export interface CustomComponentManagerArgs {
+  named: Dict<Opaque>;
   positional: Opaque[];
-  named: T;
 }
 
+export interface ManagerDelegate<ComponentInstance = Opaque> {
+  capabilities: ComponentCapabilityMask;
+  createComponent(factory: Opaque, args: CustomComponentManagerArgs): ComponentInstance;
+  updateComponent(instance: ComponentInstance, args: CustomComponentManagerArgs): void;
+  getContext(instance: ComponentInstance): Opaque;
+}
+
+export function hasAsyncLifeCycleCallbacks<ComponentInstance>(
+  delegate: ManagerDelegate<ComponentInstance>
+): delegate is ManagerDelegateWithAsyncLifeCycleCallbacks<ComponentInstance> {
+  return delegate.capabilities.asyncLifeCycleCallbacks;
+}
+
+export interface ManagerDelegateWithAsyncLifeCycleCallbacks<ComponentInstance = Opaque>
+  extends ManagerDelegate<ComponentInstance> {
+  didCreateComponent(instance: ComponentInstance): void;
+  didUpdateComponent(instance: ComponentInstance): void;
+}
+
+export function hasDestructors<ComponentInstance>(
+  delegate: ManagerDelegate<ComponentInstance>
+): delegate is ManagerDelegateWithDestructors<ComponentInstance> {
+  return delegate.capabilities.destructor;
+}
+
+export interface ManagerDelegateWithDestructors<ComponentInstance = Opaque>
+  extends ManagerDelegate<ComponentInstance> {
+  destroyComponent(instance: ComponentInstance): void;
+}
+
+export interface ComponentArguments {
+  positional: Opaque[];
+  named: Dict<Opaque>;
+}
+
+function valueForCapturedArgs(args: CapturedArguments): CustomComponentManagerArgs {
```

Comment:
TODO move into glimmer

Reply:
https://github.com/glimmerjs/glimmer-vm/pull/832 implements the interface in glimmer-vm. I'll follow up with a PR to remove the duplication once that lands.